src/Razor/test/RazorLanguage.Test/IntegrationTests/ComponentDiagnosticRazorIntegrationTest.cs

```diff
@@ -125,5 +127,68 @@ public void DirectiveAttribute_ComplexContent_ReportsError()
                 "Component attributes do not support complex content (mixed C# and markup). Attribute: '@key', text: 'Foo @Text'",
                 diagnostic.GetMessage());
         }
+
+        [Fact]
+        public void Component_StartsWithLowerCase_ReportsError()
+        {
+            // Arrange & Act
+            var generated = CompileToCSharp("lowerCase.razor", @"
+<input type=""text"" @bind=""Text"" />
+@functions {
+    public string Text { get; set; } = ""text"";
+}", throwOnFailure: false);
+
+            // Assert
+            var diagnostic = Assert.Single(generated.Diagnostics);
+            Assert.Equal("RZ10013", diagnostic.Id);
+            Assert.Equal(
+                "Component 'lowerCase' starts with a lowercase character. Component names cannot start with a lowercase character.",
+                diagnostic.GetMessage());
+        }
+
+        [Fact]
+        public void Element_DoesNotStartWithLowerCase_ReportsWarning()
+        {
+            // Arrange & Act
+            var generated = CompileToCSharp(@"
+<PossibleComponent></PossibleComponent>
```

Comment:
Can we test that `!` can be used to work around this?

Reply:
Added a test.